### DISCUSSION TITLE
Just looking for default values of function arguments

### DISCUSSION THREAD

**Discussion by liudr** [micropython/micropython#12683] [ESP32]
I was looking for the default frequency of SPI and found this. I wonder if this is how default values are given. Is this how defaults are defined in the core and ports? Is this a C++ dictionary object with .u_something as key and 500000 as value? Thanks.

https://github.com/micropython/micropython/blob/516385c4ccbb1f7fa9dbd37a44ed8ae33dcf1942/ports/esp32/machine_hw_spi.c#L452C61-L452C61

**Comment by robert-hh** [comment]
It is pure C code. The code initializes an array of mp_arg_t stuctures, which contain as the third element a union.
```
typedef union _mp_arg_val_t {
    bool u_bool;
    mp_int_t u_int;
    mp_obj_t u_obj;
    mp_rom_obj_t u_rom_obj;
} mp_arg_val_t;

typedef struct _mp_arg_t {
    uint16_t qst;
    uint16_t flags;
    mp_arg_val_t defval;
} mp_arg_t;
```

**Reply by liudr** [reply]
Thanks @robert-hh 
I guess my recent venture into python has dulled my skills in C. I wonder, in this specific case, whether it is better to define this SPI frequency in one include file so these defaults are not magic numbers in the source code.
